samples/winapi_drivelist/main.c

```diff
@@ -0,0 +1,58 @@
+#include <hal/debug.h>
+#include <hal/video.h>
+#include <nxdk/mount.h>
+#include <windows.h>
+
+int main(void)
+{
+    XVideoSetMode(640, 480, 32, REFRESH_DEFAULT);
+
+    // Mount some drives for demonstration purposes
+    BOOL ret;
+    ret = nxMountDrive('C', "\\Device\\Harddisk0\\Partition2\\");
+    if (!ret) {
+        debugPrint("Failed to mount C: drive!\n");
+        Sleep(5000);
+        return 1;
+    }
+    ret = nxMountDrive('E', "\\Device\\Harddisk0\\Partition1\\");
+    if (!ret) {
+        debugPrint("Failed to mount E: drive!\n");
+        Sleep(5000);
+        return 1;
+    }
+
+    // Retrieve drive bitmaks. Every bit represents one drive letter
+    DWORD driveBits = GetLogicalDrives();
+    if (driveBits == 0 && GetLastError() != ERROR_SUCCESS) {
+        debugPrint("Failed to retrieve drive bitmask!\n");
+        Sleep(5000);
+        return 1;
+    }
+    debugPrint("Drive bitmask: 0x%x\n\n", driveBits);
+
+
+    // Reserve buffer long enough for all possible drive strings plus null-terminator
+    char buffer[26 * 4 + 1];
+    // IMPORTANT: The size passed to GetLogicalDriveStringsA is WITHOUT the null-terminator, even though it gets written
+    DWORD charsWritten = GetLogicalDriveStringsA(sizeof(buffer)-1, buffer);
+    if (charsWritten == 0) {
+        // Additional error info can be retrieved with GetLastError()
+        debugPrint("Failed to retrieve byte strings!\n");
+        Sleep(5000);
+        return 1;
+    }
+
+    debugPrint("Drives found:\n");
+    for (char *drive = buffer; drive < buffer + charsWritten;) {
+        debugPrint("%s\n", drive);
+        while(*drive++);
```

Comment:
Wouldn't this stop on the `\0` after the first iteration (checks `'\\'` but post-increment stops at `\0`)? So on the second iteration, you'd print an empty string.

I also think that checking for the final `\0` sounds like a better idea.

Reply:
This code is tested, it does work as expected. The while loop stops when `drive` points to the byte after the `\0`.

Reply:
Oh, right. I'm an idiot. The '\\' test will still succeed, so it will fail on `\0` and go to the next letter.